### ISSUE TITLE
Single precision support?

### ISSUE THREAD

**Issue by biona001** [mateuszbaran/CovarianceEstimation.jl#82]
It seems like no matter what the type of input matrix is, the element type of `cov`'s output is always `Float64`. Is there a way to change the output so its type matches the input?

MWE (here since `z` has entries with `Float32`, I had hoped `Sigma` would have `Float32` as entries too)
```julia
using CovarianceEstimation
z = rand(Float32, 10, 10)
Sigma = CovarianceEstimation.cov(LinearShrinkage(DiagonalUnequalVariance(), :lw), z)

10×10 Symmetric{Float64, Matrix{Float64}}:
  0.0675259     0.0025485    -0.00566037  …  -0.000531266   0.00250261   -0.0033368
  0.0025485     0.0657167    -0.00378675     -0.00159049    0.00679027    0.00406199
 -0.00566037   -0.00378675    0.0877593       0.00347714    7.92699e-5   -0.00182063
  0.00658445    0.000846869  -0.00305704     -0.00115137    0.00167495   -0.00386975
  0.00644614    0.000316454  -0.00521028      0.00355928   -0.000518924   0.000131795
  0.00265943    0.00486818    0.00188952  …   0.00653186    0.00313317   -0.00358361
  0.00579734    0.00575905   -0.0109921      -0.000812353   0.00135112    0.00215032
 -0.000531266  -0.00159049    0.00347714      0.0913335    -0.00538168   -0.00330859
  0.00250261    0.00679027    7.92699e-5     -0.00538168    0.0834374     0.002173
 -0.0033368     0.00406199   -0.00182063     -0.00330859    0.002173      0.059085
```
`Statistics.cov` returns the correct type
```julia
Statistics.cov(z)

10×10 Matrix{Float32}:
  0.0750287    0.0191679   -0.042573     …  -0.00399578   0.0188228    -0.0250969
  0.0191679    0.0730186   -0.0284811       -0.0119625    0.0510713     0.0305512
 -0.042573    -0.0284811    0.0975104        0.0261524    0.000596208  -0.0136934
  0.0495233    0.00636951  -0.0229928       -0.00865971   0.0125977    -0.0291053
  0.048483     0.00238013  -0.0391878        0.0267702   -0.00390296    0.00099126
  0.0200022    0.0366148    0.0142116    …   0.0491277    0.0235653    -0.0269532
  0.0436032    0.0433152   -0.0826745       -0.00610991   0.0101621     0.016173
 -0.00399578  -0.0119625    0.0261524        0.101482    -0.0404769    -0.0248847
  0.0188228    0.0510713    0.000596208     -0.0404769    0.0927082     0.0163437
 -0.0250969    0.0305512   -0.0136934       -0.0248847    0.0163437     0.06565
```

**Comment by tlienart** [comment]
Had a quick look, don't have the time now to do a full PR yet but you might be interested in following the lead and starting a draft PR, the code structure is very repetitive so the changes suggested below should be applied everywhere, it should hopefully be straightforward

**Pb1**: automatic shrinkage that is computed (`λ` parameter, here:  https://github.com/mateuszbaran/CovarianceEstimation.jl/blob/4760963a06bb7124e6b08bca27851c1b77e43620/src/linearshrinkage.jl#L361-L365) is of type `Float64`; issue is with the `γ` which is a default Float64. Changing line 359 to `eltype(F)(κ / n)` fixes this (should be done elsewhere too) but it's not enough. 

**Pb2**: the `clamp(λ, 0.0, 1.0)` promotes to float64, changing it to `zero(T), one(T)` where `T = eltype(F)` fixes this but it's not enough (though at least here we get a float32 lambda)

**Pb3**: in `linkshrink` there is a hard `1.0`; replacing that with `one(λ)` 

Fixing all 3 problems avoids the type promotion. The same stuff should be done on all shrinkages + adding tests with small Float32 matrices to check the type remains.

**Comment by biona001** [comment]
closed by #84
